File: android_app/app/src/main/java/com/example/armenianbible/data/BibleModels.kt
package com.example.armenianbible.data

import java.util.UUID

enum class AppLanguage(val code: String, val displayName: String, val localeCode: String) {
    ARMENIAN("hy", "Հայերեն", "hy"),
    RUSSIAN("ru", "Русский", "ru"),
    ENGLISH("en", "English", "en")
}

enum class ArmenianEdition(val displayName: String) {
    ARARAT("Արարատյան (Աշխարհաբար)"),
    GRABAR("Գրաբար (Դասական)"),
    WESTERN("Արևմտահայերեն")
}

enum class AccentColorTheme(val displayName: String, val colorHex: String, val secondaryColorHex: String) {
    INDIGO("Indigo / Ինդիգո", "#6366F1", "#818CF8"),
    AMBER("Amber / Ոսկեգույն", "#F59E0B", "#FBBF24"),
    EMERALD("Emerald / Զմրուխտ", "#10B981", "#34D399"),
    ROSE("Rose / Վարդագույն", "#F43F5E", "#FB7185"),
    VIOLET("Violet / Մանուշակագույն", "#8B5CF6", "#A78BFA"),
    CYAN("Cyan / Երկնագույն", "#06B6D4", "#22D3EE")
}

enum class AIProvider(val displayName: String, val modelName: String) {
    GEMINI("Google Gemini 1.5 / 2.5", "gemini-1.5-flash"),
    CHATGPT("OpenAI ChatGPT (GPT-4o)", "gpt-4o-mini"),
    CLAUDE("Anthropic Claude 3.5", "claude-3-5-haiku-20241022")
}

enum class TextCategory(val id: String) {
    BOTH("both"),
    VERSES("verses"),
    PRAYERS("prayers"),
    FAVORITES("favorites");

    fun localizedTitle(lang: AppLanguage): String = when(this) {
        BOTH -> when(lang) {
            AppLanguage.ARMENIAN -> "Ստեղծարար & Աղոթքներ (Բոլորը)"
            AppLanguage.RUSSIAN -> "Стихи и Молитвы (Все)"
            AppLanguage.ENGLISH -> "Verses & Prayers (All)"
        }
        VERSES -> when(lang) {
            AppLanguage.ARMENIAN -> "Միայն Աստվածաշնչյան համարներ"
            AppLanguage.RUSSIAN -> "Только стихи Библии"
            AppLanguage.ENGLISH -> "Bible Verses Only"
        }
        PRAYERS -> when(lang) {
            AppLanguage.ARMENIAN -> "Միայն Նարեկացու աղոթքներ"
            AppLanguage.RUSSIAN -> "Только молитвы Нарекаци"
            AppLanguage.ENGLISH -> "St. Gregory Prayers Only"
        }
        FAVORITES -> when(lang) {
            AppLanguage.ARMENIAN -> "Միայն Ընտրյալներ"
            AppLanguage.RUSSIAN -> "Только избранное"
            AppLanguage.ENGLISH -> "Favorites Only"
        }
    }
}

enum class UpdateInterval(val id: String, val minutes: Long) {
    EVERY_HOUR("everyHour", 60),
    EVERY_6_HOURS("every6Hours", 360),
    EVERY_12_HOURS("every12Hours", 720),
    EVERY_24_HOURS("every24Hours", 1440),
    ON_SCREEN_ACTIVATION("onScreenActivation", 0),
    ON_TAP_ONLY("onTapOnly", -1);

    fun localizedTitle(lang: AppLanguage): String = when(this) {
        EVERY_HOUR -> when(lang) {
            AppLanguage.ARMENIAN -> "Ամեն 1 ժամը մեկ"
            AppLanguage.RUSSIAN -> "Каждый час"
            AppLanguage.ENGLISH -> "Every 1 hour"
        }
        EVERY_6_HOURS -> when(lang) {
            AppLanguage.ARMENIAN -> "Ամեն 6 ժամը մեկ"
            AppLanguage.RUSSIAN -> "Каждые 6 часов"
            AppLanguage.ENGLISH -> "Every 6 hours"
        }
        EVERY_12_HOURS -> when(lang) {
            AppLanguage.ARMENIAN -> "Ամեն 12 ժամը մեկ"
            AppLanguage.RUSSIAN -> "Каждые 12 часов"
            AppLanguage.ENGLISH -> "Every 12 hours"
        }
        EVERY_24_HOURS -> when(lang) {
            AppLanguage.ARMENIAN -> "Օրական 1 անգամ (24 ժամ)"
            AppLanguage.RUSSIAN -> "Раз в день (24 часа)"
            AppLanguage.ENGLISH -> "Once a day (24 hours)"
        }
        ON_SCREEN_ACTIVATION -> when(lang) {
            AppLanguage.ARMENIAN -> "Էկրանը բացելիս"
            AppLanguage.RUSSIAN -> "При открытии экрана"
            AppLanguage.ENGLISH -> "On screen unlock"
        }
        ON_TAP_ONLY -> when(lang) {
            AppLanguage.ARMENIAN -> "Միայն սեղմելիս (Ձեռքով)"
            AppLanguage.RUSSIAN -> "Только вручную"
            AppLanguage.ENGLISH -> "Manual tap only"
        }
    }
}

enum class WidgetLanguage(val id: String) {
    FOLLOW_APP("followApp"),
    ARMENIAN("armenian"),
    RUSSIAN("russian"),
    ENGLISH("english");

    fun localizedTitle(lang: AppLanguage): String = when(this) {
        FOLLOW_APP -> when(lang) {
            AppLanguage.ARMENIAN -> "Ինչպես ծրագրում (Համակարգային)"
            AppLanguage.RUSSIAN -> "Как в приложении"
            AppLanguage.ENGLISH -> "Same as App"
        }
        ARMENIAN -> "Հայերեն (Armenian)"
        RUSSIAN -> "Русский (Russian)"
        ENGLISH -> "English"
    }
}

data class BibleBook(
    val id: Int,
    val nameHy: String,
    val nameRu: String,
    val nameEn: String,
    val shortNameHy: String,
    val shortNameRu: String,
    val shortNameEn: String,
    val chaptersCount: Int
) {
    val isNewTestament: Boolean get() = id >= 40

    fun name(language: AppLanguage): String = when (language) {
        AppLanguage.ARMENIAN -> nameHy
        AppLanguage.RUSSIAN -> nameRu
        AppLanguage.ENGLISH -> nameEn
    }

    fun shortName(language: AppLanguage): String = when (language) {
        AppLanguage.ARMENIAN -> if (shortNameHy.isNotEmpty()) shortNameHy else nameHy
        AppLanguage.RUSSIAN -> if (shortNameRu.isNotEmpty()) shortNameRu else nameRu
        AppLanguage.ENGLISH -> if (shortNameEn.isNotEmpty()) shortNameEn else nameEn
    }
}

data class BibleVerseText(
    val id: Int,
    val bookId: Int,
    val chapter: Int,
    val verseNumber: Int,
    val textHy: String,
    val textRu: String,
    val textEn: String
) {
    fun text(language: AppLanguage, edition: ArmenianEdition = ArmenianEdition.ARARAT): String {
        return when (language) {
            AppLanguage.ARMENIAN -> {
                when (edition) {
                    ArmenianEdition.ARARAT -> convertToArarat(textHy)
                    ArmenianEdition.WESTERN -> convertToWestern(textHy)
                    ArmenianEdition.GRABAR -> textHy
                }
            }
            AppLanguage.RUSSIAN -> textRu
            AppLanguage.ENGLISH -> textEn
        }
    }

    fun reference(language: AppLanguage, bookName: String): String {
        return "$bookName $chapter:$verseNumber"
    }

    private fun convertToArarat(text: String): String {
        var res = text
        res = res.replace("Աստուած", "Աստված")
        res = res.replace("Աստուծոյ", "Աստծո")
        res = res.replace("Աստուծով", "Աստծով")
        res = res.replace("Յիսուս", "Հիսուս")
        res = res.replace("Յովհաննէս", "Հովհաննես")
        res = res.replace("եւ ", "և ")
        res = res.replace("եւ", "և")
        res = res.replace("ւած", "ված")
        res = res.replace("ւէ", "վե")
        res = res.replace("ւի", "վի")
        res = res.replace("ւո", "վո")
        res = res.replace("ւա", "վա")
        res = res.replace("՚", "")
        res = res.replace("'", "")
        return res
    }

    private fun convertToWestern(text: String): String {
        var res = text
        res = res.replace("Աստուծոյ", "Աստուծոյ")
        res = res.replace("Յիսուս", "Յիսուս")
        return res
    }
}

data class BibleVerse(
    val id: String = UUID.randomUUID().toString(),
    val textHy: String,
    val textRu: String,
    val textEn: String,
    val refHy: String,
    val refRu: String,
    val refEn: String,
    val isPrayer: Boolean = false
) {
    fun text(language: AppLanguage): String = when (language) {
        AppLanguage.ARMENIAN -> textHy
        AppLanguage.RUSSIAN -> textRu
        AppLanguage.ENGLISH -> textEn
    }

    fun reference(language: AppLanguage): String = when (language) {
        AppLanguage.ARMENIAN -> refHy
        AppLanguage.RUSSIAN -> refRu
        AppLanguage.ENGLISH -> refEn
    }
}

data class FavoriteItem(
    val id: String = UUID.randomUUID().toString(),
    val textHy: String,
    val textRu: String,
    val textEn: String,
    val refHy: String,
    val refRu: String,
    val refEn: String,
    val addedDateMillis: Long = System.currentTimeMillis()
)

data class ChatMessage(
    val id: String = UUID.randomUUID().toString(),
    val text: String,
    val isUser: Boolean,
    val timestamp: Long = System.currentTimeMillis()
)

data class BibleSearchResult(
    val bookId: Int,
    val bookName: String,
    val chapter: Int,
    val verseNumber: Int,
    val text: String
)

enum class VerseTag(val id: String, val icon: String, val colorHex: String) {
    FAITH("faith", "🕊️", "#38BDF8"),
    HOPE("hope", "⚓", "#10B981"),
    LOVE("love", "❤️", "#F43F5E"),
    GRIEF("grief", "🕯️", "#8B5CF6"),
    GRATITUDE("gratitude", "🙏", "#F59E0B"),
    WISDOM("wisdom", "📜", "#0EA5E9"),
    PRAYER("prayer", "✝️", "#6366F1");

    fun localizedTitle(lang: AppLanguage): String = when (this) {
        FAITH -> when (lang) {
            AppLanguage.ARMENIAN -> "Հավատք"
            AppLanguage.RUSSIAN -> "Вера"
            AppLanguage.ENGLISH -> "Faith"
        }
        HOPE -> when (lang) {
            AppLanguage.ARMENIAN -> "Հույս"
            AppLanguage.RUSSIAN -> "Надежда"
            AppLanguage.ENGLISH -> "Hope"
        }
        LOVE -> when (lang) {
            AppLanguage.ARMENIAN -> "Սեր"
            AppLanguage.RUSSIAN -> "Любовь"
            AppLanguage.ENGLISH -> "Love"
        }
        GRIEF -> when (lang) {
            AppLanguage.ARMENIAN -> "Սուգ և մխիթարություն"
            AppLanguage.RUSSIAN -> "Скорбь и утешение"
            AppLanguage.ENGLISH -> "Grief & Comfort"
        }
        GRATITUDE -> when (lang) {
            AppLanguage.ARMENIAN -> "Գոհություն"
            AppLanguage.RUSSIAN -> "Благодарность"
            AppLanguage.ENGLISH -> "Gratitude"
        }
        WISDOM -> when (lang) {
            AppLanguage.ARMENIAN -> "Իմաստություն"
            AppLanguage.RUSSIAN -> "Мудрость"
            AppLanguage.ENGLISH -> "Wisdom"
        }
        PRAYER -> when (lang) {
            AppLanguage.ARMENIAN -> "Աղոթք"
            AppLanguage.RUSSIAN -> "Молитва"
            AppLanguage.ENGLISH -> "Prayer"
        }
    }
}

data class VerseAnnotation(
    val id: String = UUID.randomUUID().toString(),
    val bookId: Int,
    val chapter: Int,
    val verseNumber: Int,
    val bookNameHy: String = "",
    val bookNameRu: String = "",
    val bookNameEn: String = "",
    val textHy: String = "",
    val textRu: String = "",
    val textEn: String = "",
    val colorHex: String? = null,
    val note: String = "",
    val tags: List<VerseTag> = emptyList(),
    val updatedAtMillis: Long = System.currentTimeMillis()
) {
    val key: String get() = "${bookId}_${chapter}_$verseNumber"

    fun text(language: AppLanguage): String = when (language) {
        AppLanguage.ARMENIAN -> textHy
        AppLanguage.RUSSIAN -> textRu
        AppLanguage.ENGLISH -> textEn
    }

    fun bookName(language: AppLanguage): String = when (language) {
        AppLanguage.ARMENIAN -> if (bookNameHy.isNotEmpty()) bookNameHy else "$bookId"
        AppLanguage.RUSSIAN -> if (bookNameRu.isNotEmpty()) bookNameRu else "$bookId"
        AppLanguage.ENGLISH -> if (bookNameEn.isNotEmpty()) bookNameEn else "$bookId"
    }

    fun reference(language: AppLanguage): String = "${bookName(language)} $chapter:$verseNumber"

    val hasContent: Boolean get() = (!colorHex.isNullOrEmpty()) || note.trim().isNotEmpty() || tags.isNotEmpty()
}

File: Sources/BibleModels.swift
import Foundation
import SwiftUI

// MARK: - Нормализация строк для сравнения без учета регистра и знаков препинания
extension String {
    var normalizedForComparison: String {
        self.lowercased()
            .components(separatedBy: CharacterSet.alphanumerics.inverted)
            .joined()
    }
}

// MARK: - Элемент Избранного (Объединенная модель для стихов дня и стихов из Библии)
struct FavoriteItem: Identifiable, Codable, Hashable {
    let id: UUID
    let isDailyVerse: Bool            // True — если это стих дня/виджета, False — если стих из Библии
    let bookId: Int?                  // ID книги в SQLite (для стихов из Библии)
    let chapter: Int?                 // Номер главы (для стихов из Библии)
    let verseNumber: Int?             // Номер стиха (для стихов из Библии)
    
    let textHy: String
    let textHyArarat: String
    let textRu: String
    let textEn: String
    let refHy: String
    let refRu: String
    let refEn: String
    
    init(
        id: UUID,
        isDailyVerse: Bool,
        bookId: Int? = nil,
        chapter: Int? = nil,
        verseNumber: Int? = nil,
        textHy: String,
        textHyArarat: String = "",
        textRu: String,
        textEn: String,
        refHy: String,
        refRu: String,
        refEn: String
    ) {
        self.id = id
        self.isDailyVerse = isDailyVerse
        self.bookId = bookId
        self.chapter = chapter
        self.verseNumber = verseNumber
        self.textHy = textHy
        self.textHyArarat = textHyArarat
        self.textRu = textRu
        self.textEn = textEn
        self.refHy = refHy
        self.refRu = refRu
        self.refEn = refEn
    }
    
    enum CodingKeys: String, CodingKey {
        case id, isDailyVerse, bookId, chapter, verseNumber
        case textHy, textHyArarat, textRu, textEn
        case refHy, refRu, refEn
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(UUID.self, forKey: .id)
        isDailyVerse = try container.decode(Bool.self, forKey: .isDailyVerse)
        bookId = try container.decodeIfPresent(Int.self, forKey: .bookId)
        chapter = try container.decodeIfPresent(Int.self, forKey: .chapter)
        verseNumber = try container.decodeIfPresent(Int.self, forKey: .verseNumber)
        textHy = try container.decode(String.self, forKey: .textHy)
        textHyArarat = try container.decodeIfPresent(String.self, forKey: .textHyArarat) ?? ""
        textRu = try container.decode(String.self, forKey: .textRu)
        textEn = try container.decode(String.self, forKey: .textEn)
        refHy = try container.decode(String.self, forKey: .refHy)
        refRu = try container.decode(String.self, forKey: .refRu)
        refEn = try container.decode(String.self, forKey: .refEn)
    }
    
    var text: String {
        let savedLang = AppGroupConstants.sharedDefaults.string(forKey: "app_language")
        let lang = savedLang ?? Bundle.main.preferredLocalizations.first ?? "hy"
        if lang.hasPrefix("ru") || lang == "russian" {
            return textRu
        } else if lang.hasPrefix("en") || lang == "english" {
            return textEn
        } else {
            if isAraratEditionSelected() && !textHyArarat.isEmpty {
                return textHyArarat
            }
            return textHy
        }
    }
    
    var reference: String {
        let savedLang = AppGroupConstants.sharedDefaults.string(forKey: "app_language")
        let lang = savedLang ?? Bundle.main.preferredLocalizations.first ?? "hy"
        if lang.hasPrefix("ru") || lang == "russian" {
            return refRu
        } else if lang.hasPrefix("en") || lang == "english" {
            return refEn
        } else {
            return refHy
        }
    }
    
    func text(for language: AppLanguage) -> String {
        switch language {
        case .armenian:
            if isAraratEditionSelected() && !textHyArarat.isEmpty {
                return textHyArarat
            }
            return textHy
        case .russian: return textRu
        case .english: return textEn
        }
    }
    
    func reference(for language: AppLanguage) -> String {
        switch language {
        case .armenian: return refHy
        case .russian: return refRu
        case .english: return refEn
        }
    }
}

// MARK: - Тематические Теги (Verse Tags)
enum VerseTag: String, CaseIterable, Identifiable, Codable, Hashable {
    case faith = "faith"
    case hope = "hope"
    case love = "love"
    case grief = "grief"
    case gratitude = "gratitude"
    case wisdom = "wisdom"
    case prayer = "prayer"
    
    var id: String { rawValue }
    
    var icon: String {
        switch self {
        case .faith: return "🕊️"
        case .hope: return "⚓"
        case .love: return "❤️"
        case .grief: return "🕯️"
        case .gratitude: return "🙏"
        case .wisdom: return "📜"
        case .prayer: return "✝️"
        }
    }
    
    var colorHex: String {
        switch self {
        case .faith: return "38BDF8"       // Sky blue
        case .hope: return "10B981"        // Emerald
        case .love: return "F43F5E"        // Rose
        case .grief: return "8B5CF6"       // Violet
        case .gratitude: return "F59E0B"   // Amber
        case .wisdom: return "0EA5E9"      // Ocean
        case .prayer: return "6366F1"      // Indigo
        }
    }
    
    func localizedTitle(for language: AppLanguage) -> String {
        switch self {
        case .faith:
            switch language {
            case .armenian: return "Հավատք"
            case .russian: return "Вера"
            case .english: return "Faith"
            }
        case .hope:
            switch language {
            case .armenian: return "Հույս"
            case .russian: return "Надежда"
            case .english: return "Hope"
            }
        case .love:
            switch language {
            case .armenian: return "Սեր"
            case .russian: return "Любовь"
            case .english: return "Love"
            }
        case .grief:
            switch language {
            case .armenian: return "Սուգ և մխիթարություն"
            case .russian: return "Скорбь и утешение"
            case .english: return "Grief & Comfort"
            }
        case .gratitude:
            switch language {
            case .armenian: return "Գոհություն"
            case .russian: return "Благодарность"
            case .english: return "Gratitude"
            }
        case .wisdom:
            switch language {
            case .armenian: return "Իմաստություն"
            case .russian: return "Мудрость"
            case .english: return "Wisdom"
            }
        case .prayer:
            switch language {
            case .armenian: return "Աղոթք"
            case .russian: return "Молитва"
            case .english: return "Prayer"
            }
        }
    }
}

// MARK: - Аннотация стиха (Цветной маркер, Заметка, Тематические теги)
struct VerseAnnotation: Identifiable, Codable, Hashable {
    let id: UUID
    let bookId: Int
    let chapter: Int
    let verseNumber: Int
    let bookNameHy: String
    let bookNameRu: String
    let bookNameEn: String
    let textHy: String
    let textHyArarat: String
    let textRu: String
    let textEn: String
    var colorHex: String?              // Выделение маркером (например, "FACC15", "4ADE80", etc.)
    var note: String                   // Текст личной мысли / размышления
    var tags: [VerseTag]               // Список назначенных тегов
    var updatedAt: Date
    
    init(
        id: UUID = UUID(),
        bookId: Int,
        chapter: Int,
        verseNumber: Int,
        bookNameHy: String = "",
        bookNameRu: String = "",
        bookNameEn: String = "",
        textHy: String,
        textHyArarat: String = "",
        textRu: String,
        textEn: String,
        colorHex: String? = nil,
        note: String = "",
        tags: [VerseTag] = [],
        updatedAt: Date = Date()
    ) {
        self.id = id
        self.bookId = bookId
        self.chapter = chapter
        self.verseNumber = verseNumber
        self.bookNameHy = bookNameHy
        self.bookNameRu = bookNameRu
        self.bookNameEn = bookNameEn
        self.textHy = textHy
        self.textHyArarat = textHyArarat
        self.textRu = textRu
        self.textEn = textEn
        self.colorHex = colorHex
        self.note = note
        self.tags = tags
        self.updatedAt = updatedAt
    }
    
    enum CodingKeys: String, CodingKey {
        case id, bookId, chapter, verseNumber
        case bookNameHy, bookNameRu, bookNameEn
        case textHy, textHyArarat, textRu, textEn
        case colorHex, note, tags, updatedAt
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(UUID.self, forKey: .id)
        bookId = try container.decode(Int.self, forKey: .bookId)
        chapter = try container.decode(Int.self, forKey: .chapter)
        verseNumber = try container.decode(Int.self, forKey: .verseNumber)
        bookNameHy = try container.decodeIfPresent(String.self, forKey: .bookNameHy) ?? ""
        bookNameRu = try container.decodeIfPresent(String.self, forKey: .bookNameRu) ?? ""
        bookNameEn = try container.decodeIfPresent(String.self, forKey: .bookNameEn) ?? ""
        textHy = try container.decode(String.self, forKey: .textHy)
        textHyArarat = try container.decodeIfPresent(String.self, forKey: .textHyArarat) ?? ""
        textRu = try container.decode(String.self, forKey: .textRu)
        textEn = try container.decode(String.self, forKey: .textEn)
        colorHex = try container.decodeIfPresent(String.self, forKey: .colorHex)
        note = try container.decodeIfPresent(String.self, forKey: .note) ?? ""
        tags = try container.decodeIfPresent([VerseTag].self, forKey: .tags) ?? []
        updatedAt = try container.decodeIfPresent(Date.self, forKey: .updatedAt) ?? Date()
    }
    
    var key: String {
        "\(bookId)_\(chapter)_\(verseNumber)"
    }
    
    func text(for language: AppLanguage) -> String {
        switch language {
        case .armenian:
            if isAraratEditionSelected() && !textHyArarat.isEmpty {
                return textHyArarat
            }
            return textHy
        case .russian: return textRu
        case .english: return textEn
        }
    }
    
    func bookName(for language: AppLanguage) -> String {
        switch language {
        case .armenian: return bookNameHy.isEmpty ? "\(bookId)" : bookNameHy
        case .russian: return bookNameRu.isEmpty ? "\(bookId)" : bookNameRu
        case .english: return bookNameEn.isEmpty ? "\(bookId)" : bookNameEn
        }
    }
    
    func reference(for language: AppLanguage) -> String {
        "\(bookName(for: language)) \(chapter):\(verseNumber)"
    }
    
    var hasContent: Bool {
        (colorHex != nil && !colorHex!.isEmpty) || !note.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty || !tags.isEmpty
    }
}


